src/ring_theory/witt_vector/basic.lean

```diff
@@ -0,0 +1,475 @@
+
+For a fixed commutative ring `R` and prime `p`,
+a Witt vector `x : 𝕎 R` is an infinite sequence `ℕ → R` of elements of `R`.
+However, the ring operations `+` and `*` are not defined in the obvious component-wise way.
+Instead, these operations are defined via certain polynomials
+using the machinery in `structure_polynomial.lean`.
+The `n`th value of the sum of two Witt vectors can depend on the 0th through `n`th values
+of the summands. This effectively simulates a “carrying” operation.
+
+## Main definitions
+
+* `witt_vector p R`: the type of `p`-typical Witt vectors with coefficients in `R`.
+* `witt_vector.coeff x n`: projects the `n`th value of the Witt vector `x`.
+* `witt_vector.map`: lifts a ring homomorphism `R →+* S` to a ring homomorphism `𝕎 R →+* 𝕎 S`.
+* `witt_vector.ghost_component n x`: evaluates the `n`th Witt polynomial
+  on the first `n` coefficients of `x`, producing a value in `R`.
+  This is a ring homomorphism.
+* `witt_vector.ghost_map`: a ring homomorphism `𝕎 R →+* (ℕ → R)`, obtained by packaging
+  all the ghost components together.
+  If `p` is invertible in `R`, then the ghost map is an equivalence,
+  which we use to define the ring operations on `𝕎 R`.
+* `witt_vector.comm_ring`: the ring structure induced by the ghost components.
+
+## Notation
+
+We use notation `𝕎 R`, entered `\bbW`, for the Witt vectors over `R`.
+
+## Implementation details
+
+As we prove that the ghost components respect the ring operations, we face a number of repetitive
+proofs. To avoid duplicating code we factor these proofs into a custom tactic, only slightly more
+powerful than a tactic macro. This tactic is not particularly useful outside of its applications
+in this file.
+
+-/
+
+noncomputable theory
+
+/-- `witt_vector p R` is the ring of `p`-typical Witt vectors over the commutative ring `R`,
+where `p` is a prime number.
+
+If `p` is invertible in `R`, this ring is isomorphic to `ℕ → R` (the product of `ℕ` copies of `R`).
+If `R` is a ring of characteristic `p`, then `witt_vector p R` is a ring of characteristic `0`.
+The canonical example is `witt_vector p (zmod p)`,
+which is isomorphic to the `p`-adic integers `ℤ_[p]`. -/
+@[nolint unused_arguments]
+def witt_vector (p : ℕ) (R : Type*) := ℕ → R
+
+variables {p : ℕ}
+
+/- We cannot make this `localized` notation, because the `p` on the RHS doesn't occur on the left
+Hiding the `p` in the notation is very convenient, so we opt for repeating the `local notation`
+in other files that use Witt vectors. -/
+local notation `𝕎` := witt_vector p -- type as `\bbW`
+
+namespace witt_vector
+
+variables (p) {R : Type*}
+
+/-- Construct a Witt vector `mk p x : 𝕎 R` from a sequence `x` of elements of `R`. -/
+def mk (x : ℕ → R) : witt_vector p R := x
+
+instance [inhabited R] : inhabited (𝕎 R) := ⟨mk p $ λ _, default R⟩
+
+/--
+`x.coeff n` is the `n`th coefficient of the Witt vector `n`.
+
+This concept does not have a standard name in the literature.
```

Comment:
Aren't these the Witt coordinates? (Witt vectors have too many coordinate systems for me to remember which is which...)

Reply:
@adamtopaz Maybe they are... I don't recall that I've ever seen a name in the literature. If you have a pointer, we'll happily change the name.

Reply:
I guess the reference for all the various coordinate systems is Hazewinkel: https://arxiv.org/abs/0804.3888
But maybe it doesn't make sense to introduce any terminology here because if and when we get big Witt vectors this might cause some confusion since the i-th coordinate in the p-typical context would correspond to the p^i-th component in the big context.

Reply:
I think I'll just leave it the way it is.